### DISCUSSION TITLE
Should we change a Register view to a shape that we can override as Orchard V1?

### DISCUSSION THREAD

**Discussion by aaronamm** [OrchardCMS/OrchardCore#12015] [Ideas]
Hello everyone. 

I am trying to custom a register page by minimize duplicated code because what I want to change are:
- Removing Username text box
- Adding First Name and Last Name text box.

I got some progress and can save first name and last name to a user entity.
![image](https://user-images.githubusercontent.com/2938310/179151543-73b4b131-9cd4-44eb-9aaf-a99f2c921169.png)

However, `I can't remove a Username text box` because it is not a shape like Orchard V1 which we can override its template by putting a new `Register.cshtml` file in a View folder of the active theme.

Here is the implement or Orchard V1:
https://github.com/OrchardCMS/Orchard/blob/dev/src/Orchard.Web/Modules/Orchard.Users/Controllers/AccountController.cs#L140

The implementation of Orchard Core CMS (V2):
https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Users/Controllers/RegistrationController.cs#L106

In Orchard Core CMS, we only return a normal ASP.NET Core MVC's view which cannot be overridden. 

Therefore, could you please give me some suggestions to override a Register.cshtml page or should we change the Register Action method to return a Register shape which uses Register.cshtml template that we can override in an active theme?

Thank you so much.

CC @Skrypt

**Comment by ns8482e** [comment]
> In Orchard Core CMS, we only return a normal ASP.NET Core MVC's view which cannot be overridden.
> 

not completely true, you can override module’s views from theme.   https://docs.orchardcore.net/en/latest/docs/reference/modules/Templates/#overriding-views

**Reply by aaronamm** [reply]
@ns8482e  I missed that document. You need a special structure not like Orchard V1 which was done in this theme.
https://github.com/OrchardCMS/TheBootstrapMachine/blob/master/Views/Register.cshtml

Thank you so much.

**Comment by aaronamm** [comment]
@ns8482e  it works and  here are what we need to do:
- Put a new Register.cshtml in `YourThemeFolder/Views/OrchardCore.Users/Registration/Register.cshtml`
- Log in as admin, go to Security > User Registration
- Check "Use site theme for registration page" option
- You may need to restart your app get all changes applied. 

![image](https://user-images.githubusercontent.com/2938310/179172463-542e0755-e2ed-48ae-9081-584dbaab4397.png)

@ns8482e putting a custom *.cshtml file in a them is great for controlling appearance of our website. 
However, `hat if we have a custom module and want to put a custom Register.cshtml in that module for encapsulating all related featurs in that module, I think we can't do it with this approach. 

I think if a Register view is a normal shape, we can create an alternate shape dynamically with `IShapeTableProvider`. and put a custom Register.cshtml in a custom module.

**Reply by ns8482e** [reply]
IMHO - MVC views and Shape fullfills two different purpose functionality and design. Register is tied to register functionality provided by the Controller hence IMHO It doesn't make sense to have it overridden in module but make sense to override in theme as far  as design is concern. 

I guess, Modules are excluded from View Location to reduce too many view locations //cc @jtkech 

However if you need you can always add custom view location, or change view result using ASP.NET Core way

**Reply by aaronamm** [reply]
@ns8482e thank you for your suggestion. 
I have a custom module name "Codesanook.Users" which adds these functionality:
- Adding a user profile to a user entity.
- Add First Name and Last Name to a registration form.
- Remove a Username text box from a registration form.

That's why I want to have `Register.cshtml` in this model for portability and self-contained.

**Reply by Skrypt** [reply]
Your custom module can always extend the user registration by implementing an EventHandler.

**Reply by aaronamm** [reply]
@Skrypt thank you for the information. Actually I have used UserEventHandler and FormHandler to extend UserProfile and customize Registration form.  `RegistrationFormHandler.cs` and `UserEventHandler` work like a charm. 

Here is the file structure in my custom module:
```
$ tree . -I "bin|obj"
.
├── Codesanook.Users.csproj
├── Controllers
│   └── HomeController.cs
├── Drivers
│   └── UserProfileDisplayDriver.cs
├── Handlers
│   ├── RegistrationFormHandler.cs
│   └── UserEventHandler.cs
├── LICENSE
├── Manifest.cs
├── Migrations.cs
├── Models
│   └── UserProfile.cs
├── README.md
├── Startup.cs
├── UserRegisterFilter.cs
├── ViewModels
│   └── UserProfileViewModel.cs
└── Views
    ├── Home
    │   └── Index.cshtml
    ├── UserProfile.Edit.cshtml
    └── _ViewImports.cshtml
```

After I have finished this custom module, I will share it to the community. I think it can be a good example for extending a user profile and customizing a registration form.
